plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalogFactory.java

```diff
@@ -0,0 +1,88 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.catalog.hms;
+
+import io.trino.plugin.base.CatalogName;
+import io.trino.plugin.hive.HdfsEnvironment;
+import io.trino.plugin.hive.HiveConfig;
+import io.trino.plugin.hive.NodeVersion;
+import io.trino.plugin.hive.metastore.HiveMetastoreFactory;
+import io.trino.plugin.iceberg.IcebergConfig;
+import io.trino.plugin.iceberg.IcebergSecurityConfig;
+import io.trino.plugin.iceberg.catalog.IcebergTableOperationsProvider;
+import io.trino.plugin.iceberg.catalog.TrinoCatalog;
+import io.trino.plugin.iceberg.catalog.TrinoCatalogFactory;
+import io.trino.spi.security.ConnectorIdentity;
+import io.trino.spi.type.TypeManager;
+
+import javax.inject.Inject;
+
+import java.util.Optional;
+
+import static io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.memoizeMetastore;
+import static io.trino.plugin.iceberg.IcebergSecurityConfig.IcebergSecurity.SYSTEM;
+import static java.util.Objects.requireNonNull;
+
+public class TrinoHiveCatalogFactory
+        implements TrinoCatalogFactory
+{
+    private final CatalogName catalogName;
+    private final HiveMetastoreFactory metastoreFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final IcebergTableOperationsProvider tableOperationsProvider;
+    private final String trinoVersion;
+    private final boolean isUniqueTableLocation;
+    private final boolean isUsingSystemSecurity;
+    private final boolean deleteSchemaLocationsFallback;
+
+    @Inject
+    public TrinoHiveCatalogFactory(
+            IcebergConfig config,
+            CatalogName catalogName,
+            HiveMetastoreFactory metastoreFactory,
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager,
+            IcebergTableOperationsProvider tableOperationsProvider,
+            NodeVersion nodeVersion,
+            IcebergSecurityConfig securityConfig,
+            HiveConfig hiveConfig)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
+        this.metastoreFactory = requireNonNull(metastoreFactory, "metastoreFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.tableOperationsProvider = requireNonNull(tableOperationsProvider, "tableOperationProvider is null");
+        this.trinoVersion = requireNonNull(nodeVersion, "trinoVersion is null").toString();
+        requireNonNull(config, "config is null");
+        this.isUniqueTableLocation = config.isUniqueTableLocation();
+        this.isUsingSystemSecurity = securityConfig.getSecuritySystem() == SYSTEM;
+        this.deleteSchemaLocationsFallback = requireNonNull(hiveConfig).isDeleteSchemaLocationsFallback();
```

Comment:
add msg to `rnn`, otherwise it's useless, as the NPE would fly anyway.

(and counter-productive in Java 17 (JEP 358))